### DISCUSSION TITLE
Zero Overhead Coderetreats 2024

### DISCUSSION THREAD

**Discussion by caichinger** [swkWien/sessions#137] [General]
# About

* Zero overhead, i.e. self-organized by those who participate
* Recurring dates serve as reminder and can be changed
* On-site and can be changed

# Locations
* [WKO EPU Forum](https://www.wko.at/wien/netzwerke/epu/meeting-raeume)
  * Opening hours: Montag - Donnerstag von 9:00 - 17:00 Uhr, Freitag von 9:00 - 14:00 Uhr
  * Address: [Straße der Wiener Wirtschaft 1, 1020 Wien](https://maps.app.goo.gl/2bDmhjEjtgtSJd739) (Praterstern)
  * By every member of Wirtschaftskammer (requires booking)
* FH Technikum Wien
  * Address:  Höchstädtplatz 6, 1200 Wien 
  * By @ReaSu 
* Anyline
  * Address: Zirkusgasse 13 / 2b, 1020 Wien
  * By @maosmurf 
*  \<add your location here\>

## Appointments

**You want to take part and the event is not yet being taken care of? Go ahead and lead the way!**

Dates:

* [x] Friday, 12.04.2024 09:00-14:00 WKO EPU Forum (reservation for "Claus Aichinger), space for up to 12 people
  * [X] Organize venue  
  * [x] Send reminder
* [x]  Saturday, 18.05.2024 09:00-14:00 (rule exception)
  * [X] Organize venue (Anyline, 1020 Vienna via @maosmurf )
  * [x] Send reminder
* [x] Friday, 14.06.2024 09:00-14:00 (WKO EPU 8 is booked on Code Mentor KG, Görge Albrecht)
  * [x] Organize venue
  * [x] Send reminder
* [x] Friday, 12.07.2024 09:00-14:00 (WKO EPU 9 booked for Peter Kofler)
  * [x] Organize venue
  * [x] Send reminder
* [ ] ~Friday, 09.08.2024 09:00-14:00~
* [ ] ~**Saturday, 14.09.2024 09:00-14:00**~
* [x] Friday, 11.10.2024 09:00-14:00
  * [x] Organize venue (FHTW)
* [x] Saturday, 09.11.2024 09:00-14:00 (Global Day)
* [ ] ~Friday, 13.12.2024 09:00-14:00~

Rules:
* Second Friday every two months starting in April until end of year.
* Second Saturday every two months starting in May until end of year.

**Comment by caichinger** [comment]
@swkWien/organisers see above for this year's Zero Overhead Code Retreat schedule.
Hoping to see many of you in course of the coming gatherings!

**Comment by caichinger** [comment]
### ZOCR 2404

Thank you for joining! :clap: 

You find today's state in above commit.

I would like that next time [we think of our template](https://github.com/swkWien/sessions/tree/main/templates) to remind us of our rules of engagement, check-in etc. I will try to do that.

I am not entirely happy with our "let's try interactive programming session", maybe I offer one in the future where I take a more active role in guiding you through the process (!= navigator role), in case you want to join. I am interesting in better understanding the relationships between IP and TDD and how the can complement each other.

Here the links to the two publications I mentioned:
* [The Curse of Recursion: Training on Generated Data Makes Models Forget](https://arxiv.org/abs/2305.17493) (Model collapse)
* [Is ChatGPT corrupting peer review? Telltale words hint at AI use](https://www.nature.com/articles/d41586-024-01051-2) (Research collapse?;)

**Comment by codecop** [comment]
Regarding Brainf*k language that you rejected. It is not nice to code it for sure but it was not created to be like that. It seems to have been created to allow the smallest compiler possible: "Compilers für den Amiga in lediglich 240 Byte zu schreiben. Brian Raiter konnte dies unterbieten, indem er – unter Verwendung von nur 171 Bytes – einen Brainfuck-Compiler für x86 Linux schrieb. Für MS-DOS gibt es einen Brainfuck-Interpreter von Bertram Felgenhauer mit einer Größe von nur 98 Bytes" So I guess the exercise would be to create the compiler not code in the language ;-) Also it seems to be proven Turing complete.

**Comment by gregorriegler** [comment]
thank you for posting the links.

> Citing [The Curse of Recursion: Training on Generated Data Makes Models Forget](https://arxiv.org/abs/2305.17493) (Model collapse)

very interresting. makes me think how this effect not only applies on llms but humans as well, like confirmation bias :)

**Comment by caichinger** [comment]
> Regarding Brainf*k language that you rejected. It is not nice to code it for sure but it was not created to be like that. It seems to have been created to allow the smallest compiler possible: "Compilers für den Amiga in lediglich 240 Byte zu schreiben. Brian Raiter konnte dies unterbieten, indem er – unter Verwendung von nur 171 Bytes – einen Brainfuck-Compiler für x86 Linux schrieb. Für MS-DOS gibt es einen Brainfuck-Interpreter von Bertram Felgenhauer mit einer Größe von nur 98 Bytes" So I guess the exercise would be to create the compiler not code in the language ;-) Also it seems to be proven Turing complete.

Interesting, I did not know that. From the perspective of [being a very limited language](https://en.wikipedia.org/wiki/Brainfuck)

> Brainfuck requires one to break commands into microscopic steps.

could be interesting to try but I am still vary about the syntax.
There is a [VS Code extension for Brainfuck](https://marketplace.visualstudio.com/items?itemName=BabyPenguin.vscode-brainfuck) though.

**Comment by caichinger** [comment]
> thank you for posting the links.
> 
> > Citing [The Curse of Recursion: Training on Generated Data Makes Models Forget](https://arxiv.org/abs/2305.17493) (Model collapse)
> 
> very interresting. makes me think how this effect not only applies on llms but humans as well, like confirmation bias :)

You are welcome :)
I too think that is pretty interesting.

---

FWIW

> [...] this effect [...]

With regards to your comparison, I am not sure though. No being an expert in any of these matters... when we follow this definition of [confirmation bias](https://catalogofbias.org/biases/confirmation-bias/) (that's a nice link as well btw):

> Confirmation bias occurs when an individual looks for and uses the information to support their own ideas or beliefs. It also means that information not supporting their ideas or beliefs is disregarded.

I would say that an LLM has no beliefs (in particular they may hallucinate) and feeding it with LLM-created content does not imho introduce certain biases but instead feeds a degenerative process that is affecting overall model performance.

However, I too think that if we continuously succumb to our own biases then yes, our model of the world collapses. In that sense I agree with you. :)

**Comment by gregorriegler** [comment]
> > thank you for posting the links.
> > > Citing [The Curse of Recursion: Training on Generated Data Makes Models Forget](https://arxiv.org/abs/2305.17493) (Model collapse)
> > 
> > 
> > very interresting. makes me think how this effect not only applies on llms but humans as well, like confirmation bias :)
> 
> You are welcome :) I too think that is pretty interesting.
> 
> FWIW
> 
> > [...] this effect [...]
> 
> With regards to your comparison, I am not sure though. No being an expert in any of these matters... when we follow this definition of [confirmation bias](https://catalogofbias.org/biases/confirmation-bias/) (that's a nice link as well btw):
> 
> > Confirmation bias occurs when an individual looks for and uses the information to support their own ideas or beliefs. It also means that information not supporting their ideas or beliefs is disregarded.
> 
> I would say that an LLM has no beliefs (in particular they may hallucinate) and feeding it with LLM-created content does not imho introduce certain biases but instead feeds a degenerative process that is affecting overall model performance.
> 
> However, I too think that if we continuously succumb to our own biases then yes, our model of the world collapses. In that sense I agree with you. :)

how would you define belief?

**Comment by maosmurf** [comment]
Hi all, as the Dojo on Saturday 18th is coming up - how and where should I send the reminder?

**Comment by caichinger** [comment]
> Hi all, as the Dojo on Saturday 18th is coming up - how and where should I send the reminder?

Thank you for hosting and reminding us! :)

For me it is up to you do the reminding part.

Options include:
* https://www.softwerkskammer.org (like https://www.softwerkskammer.org/activities/Coderetreat20240412)
  * Ping @codecop if you cannot create a new event, i.e. https://softwerkskammer.org/activities/new/
* https://www.meetup.com/.
  * @redseacomputing is taking care of meetup. Refs #115 
* https://twitter.com/swkWien
  * If you want, I can give you access. In that case, write me on Signal. Refs #110

**Comment by caichinger** [comment]
> [...]
> how would you define belief?

@gregorriegler let's continue this conversation elsewhere (e.g. at our next meeting).

**Comment by redseacomputing** [comment]
Hey there, the events were created on Softwerkskammer and Meetup, but not on coderetreat.org and twitter for now.
@caichinger @ReaSu @maosmurf @gregorriegler @goerge @codecop
cu

**Comment by adamzanyline** [comment]
Thank you @redseacomputing for announcing via swk (and meetup)!
My first tweet is out: https://x.com/swkWien/status/1788980108826398868

EDIT: adding links
* [softwerkskammer.org/activities/20240518code-retreat](https://www.softwerkskammer.org/activities/20240518code-retreat)
* [meetup.com/software-crafters-vienna/events/300927678/](https://www.meetup.com/software-crafters-vienna/events/300927678/)

**Comment by maosmurf** [comment]
Thank you @gregorriegler  and @redseacomputing  for todays session about `approvaltests` in Python using [Inline Approvals Workflow](https://github.com/approvals/ApprovalTests.Java/blob/master/approvaltests/docs/explanations/InlineApprovalsWorkflow.md).
We started via [cyber-dojo](https://beta.cyber-dojo.org/creator/choose_ltf?exercise_name=), using the ` Python 3.12, approvaltests-12.0.0 ` setup, then quickly moved to local PyCharm (and Code With me feature for faster handoff).
One can find today's code at https://github.com/maosmurf/20240518_ant_approval_2

**Comment by gregorriegler** [comment]
Today I joined the Python Approvals Mob and told them about our CodeRetreat. I explained our problem where we lost the previous results on a failing refactor, and proposed an option that would preserve the previous result. We briefly implemented it. Amazing! https://github.com/approvals/ApprovalTests.Python/releases/tag/v12.2.0

**Comment by caichinger** [comment]
Who can organize a venue for the next session on **Friday, 14.06.2024 09:00-14:00**?

At WKO there is still one free room but I have few hours left.

**Comment by goerge** [comment]
WKO EPU 8 is booked on Code Mentor KG, Görge Albrecht on 2024-06-14 09:00-14:00

HTH

**Comment by caichinger** [comment]
> WKO EPU 8 is booked on Code Mentor KG, Görge Albrecht on 2024-06-14 09:00-14:00

Thank you for taking care of that.

**Comment by ReaSu** [comment]
As you might have noticed I created the event on softwerkskammer.org and sent out an invite (thanks @goerge for the venue!), however, unfortunately I won't be able to attend.

Looking forward to July! I would really like to give interactive programming another try.

**Comment by caichinger** [comment]
> As you might have noticed I created the event on softwerkskammer.org and sent out an invite (thanks @goerge for the venue!), however, unfortunately I won't be able to attend.

Thank you for taking care of that (even though you do not have time to join yourself).

> Looking forward to July! I would really like to give interactive programming another try.

Happy to facilitate that in the future. :)

**Comment by caichinger** [comment]
I created https://pads.c3w.at/code/#/2/code/edit/IQyRDjgpy-0a2Vj6+BhDKoWR/ in case you want to document something.

See you tomorrow :)

**Comment by goerge** [comment]
@caichinger So it seems it's only the two of us. Do you still want to make it happen? If not, it's fine for me. In that case, I will cancel the WKO reservation and hopefully get my time budget back.

**Comment by goerge** [comment]
No worries. At least one "external"
participant will join. I"ll be there.

**Comment by caichinger** [comment]
> I"ll be there.

Thank you for the nice and productive session! (And not cancelling it:)

I [documented and shared](https://github.com/swkWien/sessions/commit/35aa8198ab135c5bbfde05d73f7adfb6c7e7d90d) some of our observations. Feel free to make amendments.

**Comment by ReaSu** [comment]
Is anyone up for meeting in July, or are you all on holiday?

**Comment by caichinger** [comment]
> Is anyone up for meeting in July, or are you all on holiday?

I would be in.
12, 19, 20 and 21 would work better for me though.

**Comment by codecop** [comment]
12th and 19th would work for me, Saturday is not an option. I did not want to start the discussion about moving dates, but now that Claus has started it ;-)

**Comment by ReaSu** [comment]
> 12, 19, 20 and 21 would work better for me though.

Out of these, 19th is least convenient for me.

Shall we move it to Friday, 12th?

**Please vote:**

:tada: : yes, 12th works for me
:eyes: : no, 12th doesn't work, but 13th would be fine
:slightly_frowning_face: : neither date works for me

**Comment by codecop** [comment]
@ReaSu / @goerge  are we going for 12th? I could book WKO venue?

**Comment by ReaSu** [comment]
@codecop Thank you for booking the venue!

I have created this, in case someone wants to add a topic already:
https://pads.c3w.at/code/#/2/code/edit/8vXivcYHWzWLZDqwcf3J1UrW/

Will send out an invite/reminder later today.

**Comment by codecop** [comment]
EPU Venue for 12th (this Friday) has been booked. It is "EPU 9", ask for meeting with Peter Kofler. It is available 9-14 as usual on Friday, for max 6 people. Unfortunately without windows.

**Comment by ReaSu** [comment]
Thank you!

> without windows

looking at the weather forecast, I hope there will be air conditioning! :-)

**Comment by gregorriegler** [comment]
I hope I can join. Cannot promise though.

**Comment by rolger** [comment]
I hope there is enough space available. I'll join, too
cu tomorrow

**Comment by gregorriegler** [comment]
Code from our first sessions:
https://github.com/gregorriegler/coderetreat_2024_07_12_async

**Comment by gregorriegler** [comment]
Code from our final session:
https://github.com/gregorriegler/coderetreat-2024-07-12-incremental

**Comment by caichinger** [comment]
Thank you all for contributing to this collaborative effort! :)

**Comment by codecop** [comment]
Flip with notes is in the folder https://github.com/swkWien/sessions/tree/main/2024/cr20240712

**Comment by caichinger** [comment]
### Reschedule upcoming ZOCR?

Our next CR is planned for Friday, 09.08.2024 09:00-14:00. That date does not work for me but I could to
* Saturday, 10.08.,
* Friday, 16.08. or
* Saturday, 17.08.

Those who would like to join, would you be available on one (or more) of the alternative dates?
Please share, thank you.

**Reply by codecop** [reply]
9. and 10th don't work for me. I will skip this one

**Comment by caichinger** [comment]
### Do you consent to turn this issue into a discussion?

I want to turn this issue into a discussion to make use of threads.

Are you okay with that, please :+1: otherwise :-1: .

Thank you.

**Comment by ReaSu** [comment]
Saturday, 10.8. works for me, the other dates unfortunately don't.

**Comment by ReaSu** [comment]
# Upcoming ZOCR

Dear all,

September is upon us. Shall we convene on the 14th as planned?

**Reply by maosmurf** [reply]
~Hey! On Sept 14th, I will be at https://socrates-conference.at in Linz.~
~If anyone of you is also there: would be an interesting setup for a ZOCR!~

**Reply by ReaSu** [reply]
Isn't SoCraTes in Linz one week later, 20th&21th?
You are of course right that it would be an interesting setting for a CR!

**Reply by maosmurf** [reply]
~Indeed! Therefore: I'm happy to host @ Anyline on September 14th!~

**Reply by ReaSu** [reply]
That is great, thank you!

I'll send out an invite via Softwerkskammer later this week.

**Reply by maosmurf** [reply]
@ReaSu  my bad! I just realized I will not be in Vienna on Sept 14th - therefore I will not be able to provide the location :/

**Reply by codecop** [reply]
@rolger will you join 14.9.? Maybe we meet in your office?

**Reply by ReaSu** [reply]
Oh no @maosmurf , I didn't check this before sending out the invite.
We could also meet at Technikum.

**Reply by rolger** [reply]
Sorry, I'm not available on 14th. I've an appointment with my Enkelkinder

**Comment by codecop** [comment]
@swkWien/organisers the next Small Coderetreat is scheduled for next week Friday (Friday, 11.10.2024 09:00-14:00). Is this going to happen? I can reserve a room at WKO if someone would join.

**Reply by gregorriegler** [reply]
@codecop thank you peter for the Initiative. 
This one is important as it is the last one before GDCR. Hope we can align on what we want to do for this years Event.

**Reply by caichinger** [reply]
Thank you for reaching out and offering to reserve a room.

I am in.

**Reply by codecop** [reply]
Unfortunately Friday 11th there are NO rooms available at WKO until noon (also not smaller rooms). So meeting at WKO is not an option. Which other options do we have?

**Reply by caichinger** [reply]
In case we do not find another "official" venue we can meet at my place (10 minutes by foot from U+S Hütteldorf).
(I imagine there will be suggestions, but with my offer we can plan with the date.)

**Reply by ReaSu** [reply]
Thanks Peter, for taking the lead.

I like Claus' suggestion. I could also reserve a room at Technikum, there are some available ones on that Friday.

**Reply by maosmurf** [reply]
I will not be in Vienna on that day :/

**Reply by gregorriegler** [reply]
Technikum sounds awesome to me. It's close to where I live and I'm curious how it looks from the inside.

**Reply by codecop** [reply]
So Technikum Wien it is? Would be good. I will not announce it further, as it will probably more be a discussion what to do with GDCR24 than a real CR. We might add @JDurstberger

**Reply by gregorriegler** [reply]
For Lunch I recommend 
https://vevirestaurant.at/leithastrasse1200/
It's a delicious vegan vietnamese restaurant.

**Reply by redseacomputing** [reply]
Count me in :)

**Comment by ReaSu** [comment]
I have booked room F5.11 for us. That's in the main building (the one along Dresdnerstraße), 5th floor.

**Reply by caichinger** [reply]
> I have booked room F5.11 for us

Thank you! :nerd_face: 

> [...] probably more be a discussion what to do with GDCR24 than a real CR [...]

I think so too an letting that sink it, I may not join you (and decide spontaneously) because I intended to stay out of GDCR this year.

**Comment by ReaSu** [comment]
According to the calendar there's a ZOCR tomorrow. Since noone has spoken up so far, I assume it's not taking place.  Even if it did - I couldn't attend.

**Reply by codecop** [reply]
I can't and Gregor told me he can't too. December is a busy month though
